### ISSUE TITLE
`ErrorOr` does not support null value and ignorable errors

### ISSUE THREAD

**Issue by ntkoopman** [smallrye/smallrye-graphql#2050]
I'm using the type-safe API and trying to use `ErrorOr`. In my case, the query can return a `null` value (if not found). Sometimes it also returns an error on some random (nested) field, which I would like to log.

When using `ErrorOr` I get an exception `value must not be null`, because ErrorOr uses `requireNonNull`. It also does not allow access to the value if there are any errors.

Since this class really assumes there is either a value or an error, I'm hesitant to patch it to allow null values. It would probably be better to create a new type that allows access to both without making assumptions.

**Comment by jmartisk** [comment]
> When using ErrorOr I get an exception value must not be null, because ErrorOr uses requireNonNull

I think that receiving a `null` value without an error should be allowed, I'm not sure why we disallow it. @t1 wdyt?

>  It also does not allow access to the value if there are any errors.

This is correct I think because of how GraphQL works. If there's an error on that path, there can be no (non-null) value AFAIK.

**Comment by t1** [comment]
I agree that we should patch `ErrorOr`: it's actually feasible that there is no error _and_ no value (I hadn't been thinking about this before). But in GraphQL there can never be an error _and_ a value.

**Comment by t1** [comment]
... actually you should read it as "there is no error but the value `null`"

**Comment by ntkoopman** [comment]
Yes, and since `null` is not a valid value for `Optional`/`Stream` it's a bit unclear on what the solution should be. 

It's possible to have both a value and an error, especially in the case of field errors: https://spec.graphql.org/draft/#sel-HAPHRPNBDAAADdABjoF
```gql
{
  hero(episode: $episode) {
    name
    heroFriends: friends {
      id
      name
    }
  }
}
```
```json
{
  "errors": [
    {
      "message": "Name for character with ID 1002 could not be fetched.",
      "locations": [{ "line": 6, "column": 7 }],
      "path": ["hero", "heroFriends", 1, "name"]
    }
  ],
  "data": {
    "hero": {
      "name": "R2-D2",
      "heroFriends": [
        {
          "id": "1000",
          "name": "Luke Skywalker"
        },
        {
          "id": "1002",
          "name": null
        },
        {
          "id": "1003",
          "name": "Leia Organa"
        }
      ]
    }
  }
}

```
In this case I would expect `ErrorOr` to give me access to both data and errors.

**Comment by t1** [comment]
In your example, you should place your `ErrorOr` on the `name` field. You _do_ expect to get heroes and their friends, but maybe not the name. If the error can happen everywhere, you might be better off by catching the exception.

**Comment by ntkoopman** [comment]
But I don't want an exception, I want the partial response _and_ the errors.

**Comment by t1** [comment]
> But I don't want an exception, I want the partial response and the errors.

You're right. That won't help in this case.

We've discussed about providing the partial result in the `GraphQLClientException` in #167, and split it out into #1885.

**Comment by t1** [comment]
> In this case I would like access to both data and errors. (Preferably not from a class with Or in the name though)

The `ErrorOr` javadoc states that this is theoretically possible in GraphQL, but I'm not sure if this supported, yet. And the name `ErrorOr` is okay, if you don't read it as an exclusive or 😁

**Comment by t1** [comment]
I've created a PR, but I don't see the GitHub Action running. Is this new?
